src/sentry_sync.c

```diff
@@ -91,23 +124,24 @@ worker_thread(void *data)
         } else {
             SENTRY_TRACE("executing task on worker thread");
             task->exec_func(task->data);
+            sentry__task_decref(task);
 
             sentry__mutex_lock(&bgw->task_lock);
-            bgw->first_task = task->next_task;
-            if (task == bgw->last_task) {
-                bgw->last_task = NULL;
+            if (bgw->first_task == task) {
+                bgw->first_task = task->next_task;
+                if (task == bgw->last_task) {
+                    bgw->last_task = NULL;
+                }
+                sentry__task_decref(task);
+                bgw->task_count--;
             }
-
-            bgw->task_count--;
             sentry__cond_wake(&bgw->done_signal);
             sentry__mutex_unlock(&bgw->task_lock);
-
-            if (task->cleanup_func) {
-                task->cleanup_func(task->data);
-            }
-            sentry_free(task);
         }
     }
+    if (bgw->free_self) {
+        sentry_free(bgw);
```

Comment:
Considering that the thread continues to run and send envelopes while the owning thread already timed out, this looks like the right place to free the background worker. However, this should probably free unconditionally, essentially calling a `decref` on the state.